receiver/scraperhelper/settings.go

```diff
@@ -0,0 +1,58 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package scraperhelper // import "go.opentelemetry.io/collector/receiver/scraperhelper"
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"go.uber.org/multierr"
+
+	"go.opentelemetry.io/collector/component"
+)
+
+var (
+	errNonPositiveInterval    = errors.New("requires positive value")
+	errTimeoutExceedsInterval = errors.New("timeout value exceeds collection interval")
+)
+
+// ScraperControllerSettings defines common settings for a scraper controller
+// configuration. Scraper controller receivers can embed this struct, instead
+// of receiver.Settings, and extend it with more fields if needed.
+type ScraperControllerSettings struct {
+	// CollectionInterval sets the how frequently the scraper
+	// should be called and used as the context timeout
+	// to ensure that scrapers don't exceed the interval.
+	CollectionInterval time.Duration `mapstructure:"collection_interval"`
+	// InitialDelay sets the initial start delay for the scraper,
+	// any non positive value is assumed to be immediately.
+	InitialDelay time.Duration `mapstructure:"initial_delay"`
+	// Timeout is used to set scraper's context deadline, it must be within
+	// the range of (0, CollectionInterval]
+	Timeout time.Duration `mapstructure:"timeout"`
+}
+
+// NewDefaultScraperControllerSettings returns default scraper controller
+// settings with a collection interval of one minute.
+func NewDefaultScraperControllerSettings(component.Type) ScraperControllerSettings {
+	return ScraperControllerSettings{
+		CollectionInterval: time.Minute,
+		InitialDelay:       time.Second,
+		Timeout:            time.Minute,
+	}
+}
+
+func (set *ScraperControllerSettings) Validate() (errs error) {
+	if set.CollectionInterval <= 0 {
+		errs = multierr.Append(errs, fmt.Errorf(`"collection_interval": %w`, errNonPositiveInterval))
+	}
+	if set.Timeout <= 0 {
+		errs = multierr.Append(errs, fmt.Errorf(`"timeout": %w`, errNonPositiveInterval))
+	}
+	if set.Timeout > set.CollectionInterval {
+		errs = multierr.Append(errs, errTimeoutExceedsInterval)
```

Comment:
This means that when a user of a scraping component wants to decrease the `collection_interval`, they will also _have_ to decrease the `timeout` accordingly. For example this configuration will not work:

```yaml
receivers:
  hostmetrics:
    collection_interval: 10s
    scrapers:
      memory:
```

Because user will also need to set the `timeout` accordingly:

```yaml
receivers:
  hostmetrics:
    collection_interval: 10s
    timeout: 10s
    scrapers:
      memory:
```

I'm not in favor of such an API.

How about instead of erroring out when the timeout is higher than the collection interval, we only log a warning and adjust the timeout to collection interval automatically? Something like "warning: `timeout` is 1m, which is higher than `collection_interval` 10s, resetting `timeout` to 10s" or similar. What do you think?

Reply:
Fail fast should be the right approach, especially because this can be caught in validating configuration. I am 100% in favor of erroring here instead of doing magic.

Reply:
I can understand that the "magic" approach has its disadvantages. But I still don't think it's a good API design to force a user to change two properties (`collection_interval` and `timeout`) when they only want to change one (`collection_interval`).

Also in the current state, this change that is an `enhancement`, will be a breaking change in contrib, as it will break existing user configurations for scraping components. This configuration:

```yaml
receivers:
  hostmetrics:
    collection_interval: 10s
    scrapers:
      memory:
```

is valid in current versions of contrib, but it will be invalid after this change is included in contrib. Users will need to supplement their configurations with `timeout` property reflecting the value of `collection_interval`.

I believe we should try harder to find a solution that allows this feature to exist without troubling users.

Perhaps another approach is to have a special (default) value that means "make timeout equal to collection interval"?

Reply:
> Perhaps another approach is to have a special (default) value that means "make timeout equal to collection interval"?

I would prefer not to have special values defining if a timeout should default to the collection interval.

I think a reasonable alternative is if the zero value is used (since zero is not a valid timeout), it can use the default interval.

I believe that satisfies the requirement of not having a scraper exceed its collection interval and listening to the timeout?

Does that work for you @astencel-sumo ?